Add parallel batch mode to scripted tests

jvican opened the pull request:
This pull request has two goals:
  
* Speed up scripted tests by allowing them to run in parallel.
* Allow scripted to run scripted tests by reusing the same sbt instance.
  
However, these tasks ended up not being easy. This is why this PR also addresses other issues, like the ones described in #3160.
  
The integrity of the tests has not been changed, except for one: `auto-plugins`, which still tests all the behaviour of plugins but uncovered https://github.com/sbt/sbt/issues/3164. As a result, I've done a temporary fix that accounts for the fact that `reload` in the batch mode will modify the value being tested by one.
  
This PR also deletes some deprecated methods in scripted and adjusts the code (both for scripted sbt and scripted plugin) accordingly.

jvican committed:
Delete scripted deprecations & CompatibilityLevel

We don't have to care about migrations in these removals because no one
should be depending on these methods directly. People consume scripted
via the sbt plugin.

This commit also removes the unused `CompatibilityLevel` which is
completely outdated, making reference to Scala 2.9 et al.

jvican committed:
Run scripted tests in parallel

Add binary-compatible methods to run scripted tests in parallel using
parallel collections. Parallel collections are not tuned and a default
of 4 cores is used for the `ForkJoinPool`. This can be a matter of
experimentation in the future but for now I prefer to stay with it.

The non-parallel methods are not affected by this change.

The changes to the scripted plugin have not been used, only the sbt
scripted plugin is using `runInParallel`. In the future, this change can
also happen so that plugin authors can benefit from parallel sbt
testing. This is not a priority for me now :)

dwijnand reviewed (dismissed):
LGTM.

But I expect Travis CI to fail because those deprecations are more tricky than just removing them - that's why I had left them alone for now.

You might want to just drop those hunks to get this green and merged.

dwijnand commented:
Oh, and well done, thank you very much for this!

jvican committed:
Abstract over `runAll` and `runInParallel`

This is the best we can do without using structural types.

jvican committed:
Deduplicate `scriptedTest` and remove exceptions

The following commit minimises the amount of exceptions that leak
through our internal signatures. It thus changes the return type
signature of `scriptedTest` to return an `Option` instead of `Unit`.

At the call-site of `scriptedTest`, we also remove the code to handle
the exceptions thrown by it.

eed3si9n reviewed (changes requested):
scripted forks a new instance of sbt, so I don't know if this would work on Travis CI containers due to limited memory.

jvican commented:
Is there a limitation in Travis that I'm now aware of? I don't see why this should fail. @eed3si9n

jvican commented:
> scripted forks a new instance of sbt, so I don't know if this would work on Travis CI containers due to limited memory.
  
@eed3si9n You're right. This brings me to the second improvement that I wanted to make: reusing sbt instances to run things in parallel.
  
This is a shame since tests are running much faster locally. The overhead of sbt scripted tests is mainly I/O -- this is why parallelizing them yields profitable gains.
  
Honestly, I think that Travis is not enough for sbt. I would be happy to put sbt/sbt under the Scala Center CI, where we have 64 cores and 14GB of memory. Would you be happy with it? /cc @dwijnand

jvican committed:
Reduce memory of scripted tests + extra

* Remove oversight in error message
* Up the memory limit per scripted instance

jvican commented:
I have pushed a commit that reduces the memory for every scripted test, in the hope that this makes Travis happy... let's see.

jvican commented:
No luck, unfortunately. I'll try to make my next improvement to see if that helps us get out of this situation. Otherwise, I would suggest to change CIs, at this point it's clear that Travis is a blocker...

eed3si9n commented on scripted/sbt/src/main/scala/sbt/test/SbtHandler.scala:
```diff
@@ -80,7 +80,10 @@ final class SbtHandler(directory: File,
     val thread = new Thread() { override def run() = { p.exitValue(); server.close() } }
```
```
[error] x compiler-project / inc-pickled-refinement compiler-project / inc-pickled-refinement
[error]  Cause of test exception: {line 7}  Command failed: Remote sbt initialization failed
```

The forked sbt on the other side is failing to get up, so you might need more println around here where the Process is being spawned.

jvican replied:
Thanks for the tip.

I think the issue here is memory since all these tests are passing locally. Can we discuss migrating CI to a more powerful machine, either Drone (free) or paying for a more powerful Travis?
  
We need to find a good way to move forward.

eed3si9n replied:
We can potentially discuss it, but I'm relatively happy with Travis CI as we get 5 workers for free. I think that's pretty good ROI.

jvican replied:
You get 5 workers for free but the turnaround time is at minimum 45 minutes for every commit you do, and there's usually a long queue. We need more powerful alternatives. I will not insist on using our free CI, but can you please consider paying a pro account in Travis to increase the CI resources?

Being honest here, contributing to sbt is less than an ideal experience. I've tried solving it by parallelizing the testing engine, but even if that is blocked by Travis resources, I don't know what else to do. In my view, there is no other way to move forward than increasing the resources.

jvican committed:
Use `ConsoleReporter` if no reporter is found

This change is necessary in the cases where we have global
initialization issues that have no position, like:

```
[info] [error] scala.reflect.internal.MissingRequirementError: object scala in compiler mirror not found.
```

Before, it was failing with a `sys.error` exception. Now we will report
these issues with a console reporter that is not meant to be
thread-safe.

jvican committed:
Fix #3160: Make actions/task-cancel compile

jvican committed:
Fix #3160: Disable sha-conflict

This test was always failing but was marked as passed by the previous
scripted framework because the test files were only checking the failure
of tests.

The test case depends on bintray, which could never have been resolved
for sbt 1.0.0-SNAPSHOT. This resolution error was the one "hidden".

jvican committed:
Fix #3160: Don't use val extractor in build

See comment in https://github.com/sbt/sbt/issues/3160.

jvican committed:
Remove `MaxPermSize` since sbt depends on JDK8

FTR:

```
[error] OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=256M; support was removed in 8.0
```

jvican committed:
Fix #3161: Don't fail if group is not present

This covers the case described in #1361 plus an esoteric case too.
Parsers are initialized when sbt starts and text completions use the
knowledge at that time. If we happen to move a new file that should
influence the behaviour of the tab completion, that file will not be
picked up. Before, since we were throwing an error, the previous case
that would no produce any tab completion would also fail, though the
files are legitimately in the correct place.

jvican committed:
Add batch mode execution to scripted

For that, we:

* Change the existing infrastructure to recycle as much code as
  possible.
* Use `BatchScriptRunner` since `ScriptRunner` is too restrictive to
  programmatically control the underlying sbt servers.
* Unify `TestRunner` to the more general way for both batch and
  non-batch modes.

jvican committed:
Run pre hook only after checking disabled tests

jvican committed:
Use scripted hook to run reload

This saves us from duplicating error handling logic to correctly manage
`pending` tests whose projects don't even load (and make our reload)
fail.

jvican committed:
Fix compile tests using `checkIterations`

Those tests using `checkIterations` were not working because an
invariant was working: the first iterations number is 0. As we reuse
sbt for running several tests, this is no longer true, so this commit
makes sure that they are updated with the pertinent infrastructure to
compute the performed iterations from the latest known number of
iterations.

jvican committed:
Fix java options for scripted

* Remove MaxPermSize from another scripted opts
* Reduce memory of sbt host to 1g instead of 2g
* Add Xms java options to scripted
* Enable parallelism with 512M for sbt tests
* Parallelism + batch mode

jvican committed:
Fix fancy uses of analysis

Clean up first analysis to make sure that test utilities using
information from analysis have correct information that does not collide
with the compilation analysis of the previous runs (batch tests).

jvican committed:
Fix #3160: Newline after comment line

jvican committed:
Fix more tests using compiler iterations

jvican committed:
Fix #3610: project/old-ops

`compile` does nothing in this test, what we need to do is `reload`.

jvican committed:
Fix project/binary-plugin: Reload was implied

The first `>` that appears kickstarts the execution of sbt, so it
implies a reload. This is no longer true for the batch test execution.

jvican committed:
Set name of root project if no name is known

This commit makes sure that between every run in batch mode the local
root project name is updated if no name is set.

This fixes project/generated-root-no-project that was assuming that the
root directory name was `generated-root-no-project`. This invariant does
not hold anymore with the batch-mode. Now one sbt dir is shared for lots
of scripted tests.

jvican committed:
Try travis with 3 cores and 725M per instance

jvican committed:
Reboot to ditch state produced by scripted reload

jvican committed:
Create a compiler plugin for scripted

The compiler plugin is more appropriate than running a series of
commands because of the effects of `set`.

jvican committed:
Fix auto-plugins temporarily until #3163 is fixed

Unfortunately, I think this uncovers a bug in duplicated computation of
settings in the `ThisBuild` and `Global` axis.

jvican committed:
Fix string-to-command to ignore existing commands

The scripted batch executor injects a command and the operation
performed in this test checks all the commands, assuming only `noop` is
declared.

jvican committed:
Handle empty case in `labelsAndDirs`

jvican committed:
Up memory for every scripted instance

jvican committed:
Try another travis structure + build tweaks

I've tried to put together some scripted tests to remove the overhead of
compiling the whole sbt, which is around 3 minutes every time.

This new structure *should* make the scripted tests run faster.

Aside from this, we do some more tweaks:

* Increase memory.
* Fork processes to compile and run (to see if it makes a difference).
* Pass in the server flag to sbt.

jvican commented:
@dwijnand @eed3si9n This is ready for review. Tests are passing and description is updated. I will report on the benefits brought by this PR tomorrow.

jvican commented:
Note to myself: I have yet to update to latest librarymanagement, but I'll do that in another PR.

eed3si9n commented on main/src/main/scala/sbt/internal/parser/SbtParser.scala:
```diff
@@ -99,14 +98,15 @@ private[sbt] object SbtParser {
-    import scala.reflect.internal.util.NoPosition
     val options = "-cp" :: s"$defaultClasspath" :: "-Yrangepos" :: Nil
-    val reportError = (msg: String) => globalReporter.error(NoPosition, msg)
+    val reportError = (msg: String) => System.err.println(msg)
```
We get into trouble when we start printlning directly. I am guessing this is to surface scripted failures, but would this have negative impact to normal usage?

jvican replied:
No, not at all. In fact, this should never happen. It happened only in this case because I was fiddling with the classpath of the compiler. But this should not affect normal users.

eed3si9n commented on scripted/sbt/src/main/scala/sbt/test/ScriptedTests.scala:
```diff
@@ -8,120 +8,277 @@ package test
+
+          // Run the test and delete files (except global that holds local scala jars)
+          val result = runOrHandleDisabled(label, tempTestDir, runTest, buffer)
+          IO.delete(tempTestDir.*("*" -- "global").get)
```
This wipes out the forked instances temp directory? So you're saving time by not resolving sbt 1.0.0-SNAPSHOT from the second batch onwards?

jvican replied:
Yes, this wipes out all the content of the temp directory except the global dir holding the scala jars. The benefits of running scripted in batch more are several, though -- for instance, you always keep a hot sbt instance around.

eed3si9n reviewed (commented):
Overall looks pretty good.
Posted a few minor questions for clarification.

jvican committed:
Split project and aggregate project-load

jvican commented:
After playing around with lots of options, this PR is now finished.
  
**Improvements**:
  
* Batch mode execution.
* Parallel execution (by default 1 sbt instance because Travis doesn't allow more).
* Forked sbt compilation and scripted run.
* Tests in #1360 fixed.
  
It is a large PR (in number of commits), but line changes are small. I suggest reviewing all at once, and if there's a concrete question you can check if the commit message that introduced a change explains it.
  
As a result, and even with only one sbt instance running scripted tests, this PR speeds up the scripted tests by almost 2x.
* 36 minutes: https://travis-ci.org/scalacenter/sbt/builds/228317280.
* 61 minutes: https://travis-ci.org/sbt/sbt/builds/224407414.
  
I believe we may win a little bit more of a time when util-X11 is merged, which brings parallel ivy downloads.
  
In the future, when performance issues and memory leaks are addressed in sbt, I will come back to scripted to try to run 2 sbt instances at a time.
  
In the meanwhile, if you wanna run scripted locally, I suggest that you change the number of sbt instances [here](https://github.com/sbt/sbt/pull/3151/files#diff-68598a656c879eaa71bb207678c547cfR347). It does bring a speedup.

Duhemm commented on sbt/src/sbt-test/project/auto-plugins/build.sbt:
```diff
@@ -52,9 +52,9 @@ check := {
+	same(buildValue, "build 1", "demo in ThisBuild") // this is temporary, should be 0 until # is fixed
 	val globalValue = (demo in Global).value
-	same(globalValue, "global 0", "demo in Global")
+	same(globalValue, "global 1", "demo in Global") // this is temporary, should be 0 until # is fixed
```
Until what is fixed?

jvican replied:
#3164

Duhemm commented on sbt/src/sbt-test/project/old-ops/test:
```diff
@@ -1,17 +1,18 @@
 
 $ copy-file changes/taskAppendN/build.sbt build.sbt
--> compile
+-> reload
```
So scripted was starting up a whole new sbt between every command???

jvican replied:
What I meant is that `sbt` was initialized lazily, only when `>` statements were invoked. `compile` here implied initializing sbt.

Duhemm commented on sbt/src/sbt-test/apiinfo/unstable-existential-names/build.sbt:
```diff
@@ -1,12 +1,26 @@
+checkIterations := {
+  val expected: Int = (Space ~> NatBasic).parsed
+  val actual: Int = ((compile in Compile).value match { case a: Analysis => a.compilations.allCompilations.size }) - CompileState.previousIterations
+  assert(expected == actual, s"Expected $expected compilations, got $actual")
```
Shouldn't this test be in zinc's suite?

jvican replied:
I have to port these improvements to Zinc, indeed! The issue is that this has to be included in every build.sbt. This is why we have to put it in here.

Duhemm commented on sbt/src/sbt-test/compiler-project/inc-pickled-existential/build.sbt:
```diff
@@ -1,15 +1,28 @@
+checkIterations := {
+  val expected: Int = (Space ~> NatBasic).parsed
+  val actual: Int = ((compile in Compile).value match { case a: Analysis => a.compilations.allCompilations.size }) - CompileState.previousIterations
+  assert(expected == actual, s"Expected $expected compilations, got $actual")
```
This one should be moved to, shouldn't it? I think zinc's scripted supports this kind of assertion.

There are certainly other tests that could be moved over to zinc. Is there a reason to keep them here?

jvican replied:
All these tests should be moved to Zinc yes. But I'd like to do that in the future, I don't have enough cycles for it now.

Duhemm commented on sbt/src/sbt-test/source-dependencies/trait-member-modified/build.sbt:
```diff
@@ -22,7 +32,7 @@ TaskKey[Unit]("checkCompilations") := {
     assert(recompiledFiles(iteration) == files, "%s != %s".format(recompiledFiles(iteration), files))
   }
-  assert(allCompilations.size == 2)
+  assert(allCompilations.size == 2, s"All compilations is ${allCompilations.size}")
```
Maybe say that we expected 2?

Duhemm reviewed (changes requested):
Awesome improvements, very impressive!

 - Why has the test `project-load/sha-conflict` moved from `pending` to `disabled`?
 - Travis CI kills a build if it doesn't produce any output for 10 minutes. It appears that people circumvent that issue by disabling log buffering (hopefully they'll never have to do that again, thanks to that PR 🎉). Would it be possible to have scripted periodically write something next to the test name, to show that it's not dead? Something like
```
Running run / hello-world .............. done!
```

jvican commented on sbt/src/sbt-test/project/auto-plugins/build.sbt:
```diff
@@ -52,9 +52,9 @@ check := {
+	same(buildValue, "build 1", "demo in ThisBuild") // this is temporary, should be 0 until # is fixed
 	val globalValue = (demo in Global).value
-	same(globalValue, "global 0", "demo in Global")
+	same(globalValue, "global 1", "demo in Global") // this is temporary, should be 0 until # is fixed
```
#1364

eed3si9n commented:
> It is a large PR (in number of commits), but line changes are small. I suggest reviewing all at once, and if there's a concrete question you can check if the commit message that introduced a change explains it.

Once the review is done, could we squash them into 1 or several number of commits for future debugging purpose?

jvican commented:
@eed3si9n Honestly, I prefer not to. They have been done this way so that they are easy to reason about, and each one implements a granular change that may be useful for the future. That being said, I would like to squash those commits that have to do with travis options and "disable parallelism".

jvican commented:
> Travis CI kills a build if it doesn't produce any output for 10 minutes. It appears that people circumvent that issue by disabling log buffering (hopefully they'll never have to do that again, thanks to that PR 🎉). Would it be possible to have scripted periodically write something next to the test name, to show that it's not dead?
  
This would be an excellent issue for a Scala Center spree, thanks for the idea!

jvican commented:
Can we try to get this in? When reviews are done, I'll squash the pertinent commits. I'd like to get this in first before making more PRs.

jvican commented:
I'm squashing the pertinent commits and rebasing to prepare for the merge. Will report when this is ready.

jvican commented:
Can someone reload this failed build instance? https://travis-ci.org/sbt/sbt/jobs/230899635 The culprit is https://github.com/sbt/sbt/issues/3170.
  
The rest of the tests are passing.

jvican commented:
Tests are passing.

muuki88 commented:
@jvican great work! Sbt-native-packager scripted test run for almost 40 minutes. This will be a huge development experience improvement 😍😍
